### PR TITLE
Added initial dataset interface and EDA methods

### PR DESCRIPTION
**Pull request by werag55**

### CODE REVIEW

File: src/eda/eda.py [hunk 1]
```diff
@@ -0,0 +1,53 @@
+"""Module providing functions for Exploratory Data Analysis."""
```
Comment (mytkom) [inline src/eda/eda.py:1]: It seems nice to have every EDA plot in such a method, but I have one concern. There are a lot of possible parameters to add to every method and it doesn't seem easy to determine which of them are important. For sure I think that the `figsize` parameter should be present in all method definitions.

File: src/eda/eda.py [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+    plt.show()
+
+def plot_target_distribution(dataset):
+    """Plot the distribution of the target variable."""
+    plt.figure(figsize=(8, 6))
+    sns.countplot(x=dataset.labels)
+    plt.title("Target Variable Distribution")
+    plt.xlabel("Class")
+    plt.ylabel("Count")
+    plt.show()
```
Comment (mytkom) [inline src/eda/eda.py]: I think `plt.show()` makes this method usable only in the single context: "jupyter notebook". Maybe it should return a `Figure` object, which can then be, e.g. saved to a file or shown to Jupyter output. The same argument applies to other plotting methods.

File: src/eda/eda.py [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+    """Plot pairwise relationships between features (sampled for performance)."""
+    sample_data = dataset.data.sample(sample_size)
+    sample_data['label'] = dataset.labels[sample_data.index]
+    sns.pairplot(sample_data, hue='label', palette='viridis')
+    plt.suptitle("Pairwise Feature Relationships", y=1.02)
+    plt.show()
+
+def plot_tsne(dataset, perplexity=30):
+    """Plot t-SNE visualization of the dataset."""
+    tsne = TSNE(n_components=2, perplexity=perplexity, random_state=42)
```
Comment (mytkom) [inline src/eda/eda.py]: This random state integer should be passed by a parameter or set as a shared project constant.

File: tests/data/test_data_loader.py [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+    data_interface = DataInterface(SyntheticDataLoader(p, n, d, g))
+
+    assert data_interface.data.data.shape == (n, d)
+    assert len(data_interface.data.labels) == n
+    assert set(data_interface.data.labels).issubset({0, 1})
+    mean_class_0 = data_interface.data.data[data_interface.data.labels == 0].mean()
+    mean_class_1 = data_interface.data.data[data_interface.data.labels == 1].mean()
+    assert np.allclose(mean_class_0, np.zeros(d), atol=0.1)
+    expected_mean_class_1 = np.array([1 / (k + 1) for k in range(d)])
+    assert np.allclose(mean_class_1, expected_mean_class_1, atol=0.1)
```
Comment (mytkom) [inline tests/data/test_data_loader.py:30]: Can we be sure that this test always passes? Is `atol` tolerance of the mean estimation? I'm little bit lost

File: tests/data/test_data_loader.py [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+    assert data_interface.data.data is not None
+    assert data_interface.data.labels is not None
+    assert isinstance(data_interface.data, DatasetInterface)
+    assert isinstance(data_interface.data.data, pd.DataFrame)
+    assert isinstance(data_interface.data.labels, pd.Series)
+
+def test_synthetic_data_loader():
+    """Test synthetic data generation."""
+    p, n, d, g = 0.5, 10000, 10, 0.5
+    data_interface = DataInterface(SyntheticDataLoader(p, n, d, g))
```
Comment (mytkom) [inline tests/data/test_data_loader.py:21]: I really like wrapping DataLoaders in this interface, it is pretty tidy.

File: tests/data/test_dataset_interface.py [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+
+def test_convert2binary():
+    """Test converting labels to binary values."""
+    data_interface = DataInterface()
+    data_interface.data.labels = pd.Series([0, 1, 2, 0])
+
+    data_interface.convert2binary()
+
+    assert set(data_interface.data.labels.unique()) == {0, 1}
+    tm.assert_series_equal(data_interface.data.labels, pd.Series([1, 0, 0, 1]))
```
Comment (mytkom) [inline tests/data/test_dataset_interface.py]: So we map instances of input class `0` to output class `1` and instances of any other class to output class `0`. I think we should just remove samples with labels of any other class than passed to `convert2binary(in_label_0, in_label_1)` or something. What do you think?

File: tests/data/test_dataset_interface.py [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+    data_interface = DataInterface()
+    data_interface.data.data = pd.DataFrame({
+        'feature1': [1, 2, np.nan, 4],
+        'feature2': [5, np.nan, 7, 8]
+    })
+    data_interface.data.labels = pd.Series([0, 1, 0, 1])
+
+    data_interface.handle_missing_values(strategy='drop')
+
+    assert data_interface.data.data.shape[0] == 2
```
Comment (mytkom) [inline tests/data/test_dataset_interface.py:36]: I am unsure if this `data_interface.data.data.shape[0]` returns the number of features or samples. Please check it

File: src/data/data_loader.py [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+
+
+class SyntheticDataLoader(DataLoader):
+    """Synthetic data creator."""
+
+    def __init__(self, p, n, d, g):
+        self.p = p
+        self.n = n
+        self.d = d
+        self.g = g
```
Comment (mytkom) [inline src/data/data_loader.py]: Please describe one letter variables in docstring

File: src/data/data_loader.py [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+            for j in range(self.d):
+                cov_matrix[i, j] = self.g ** abs(i - j)
+
+        data = np.zeros((self.n, self.d))
+        for i in range(self.n):
+            if labels[i] == 0:
+                mean = np.zeros(self.d)
+            else:
+                mean = np.array([1 / (k + 1) for k in range(self.d)])
+            data[i, :] = multivariate_normal.rvs(mean=mean, cov=cov_matrix)
```
Comment (mytkom) [inline src/data/data_loader.py]: What is the difference between `np.random.multivariate_normal()` and this scipy's `rvs()` method?

File: src/data/dataset_interface.py [hunk 10]
```diff
@@ -0,0 +1,155 @@
...
+
+    def __init__(self, data_loader=None):
+        self.data = DatasetInterface()
+        if data_loader:
+            self.data.data, self.data.labels = data_loader.load_data()
+        self.train_data = DatasetInterface()
+        self.val_data = DatasetInterface()
+        self.test_data = DatasetInterface()
+
+    def preprocess_data(self, *, missing_values_strategy='mean'):
```
Comment (mytkom) [inline src/data/dataset_interface.py]: what `*` means as parameter?

File: src/data/dataset_interface.py [hunk 11]
```diff
@@ -0,0 +1,155 @@
...
+        Preprocess the dataset.
+        """
+        self.handle_missing_values(strategy=missing_values_strategy)
+        self.encode_labels()
+        self.convert2binary()
+        self.remove_correlated_features()
+
+        num_observations = len(self.data.data)
+        num_features = self.data.data.shape[1]
+        required_features = int(0.5 * num_observations)
```
Comment (mytkom) [inline src/data/dataset_interface.py]: maybe this `0.5` ratio should be passed as a parameter or at least be a class constant variable.

**Review by mytkom — COMMENTED** [review]
I really like your interface, and I have just a few questions (some just lack Python knowledge). We can discuss it at the meeting later.

**Review by mytkom — APPROVED** [review]
